### PR TITLE
DDO-2937 workflow call for gha terra-ui ci tests

### PR DESCRIPTION
**Pull request by jyang-broad**
## Summary of changes:
Adds a new workflow_call github action that other repos can call to run tests

### What
- adds a github action
- does not update or edit existing terra-ui code.

### Why
- allows for CI integration using GHA

### Testing strategy
- https://github.com/broadinstitute/monkey/actions/runs/5491174959 test runs from a private repo, to confirm no leakage of secrets.

- Example Test call:
```
# Test workflow to run terra-ui tests from GHA
name: Test Run TerraUI Workflow Call
on:
  pull_request:
    paths-ignore:
      - 'README.md'
  push:
    branches:
      - main
    paths-ignore:
      - 'README.md'
jobs:
  
  run-terraui-integration-tests:
    permissions:
      contents: 'read'
    # Put new Sam version in Broad dev environment
    uses: ./.github/workflows/run-integration-tests.yaml
    with:
      test_url: 'https://bvdp-saturn-dev.appspot.com/'
      environment: 'dev'
    secrets:
      LYLE_CLIENT_EMAIL: ${{ secrets.LYLE_CLIENT_EMAIL }}
      LYLE_PRIVATE_KEY: ${{ secrets.LYLE_PRIVATE_KEY }}
      FIRECLOUD_ACCOUNT_CLIENT_EMAIL: ${{ secrets.FIRECLOUD_ACCOUNT_CLIENT_EMAIL }}
      FIRECLOUD_ACCOUNT_PRIVATE_KEY: ${{ secrets.FIRECLOUD_ACCOUNT_PRIVATE_KEY }}
```

### CODE REVIEW

File: .github/workflows/run-integration-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+          LYLE_SA_KEY=$(jq --null-input \
+          --arg client_email "${{ secrets.LYLE_CLIENT_EMAIL }}" \
+          --arg key"${{ secrets.LYLE_PRIVATE_KEY }}" \
+          '{"client_email": $client_email, "key": $key}') >> $GITHUB_ENV
+          
+          echo \
+          TERRA_SA_KEY=$(jq --null-input \
+          --arg client_email "${{ secrets.FIRECLOUD_ACCOUNT_CLIENT_EMAIL }}" \
+          --arg key "${{ secrets.FIRECLOUD_ACCOUNT_PRIVATE_KEY }}" \
+          '{"client_email": $client_email, "key": $key}') >> $GITHUB_ENV
```
Comment (nawatts) [inline .github/workflows/run-integration-tests.yaml:116]: Would it be simpler to have each SA key stored as a secret instead of reconstructing them from two secrets that have to be coordinated?

File: .github/workflows/run-integration-tests.yaml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+          yarn test ${{ inputs.tests_to_run }}
+
+      - name: 'Upload Test Artifacts'
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: terra-ui-logs-${{ github.run_id }}
+          path: |
+            /tmp/test-results
+            ./terraui/integration-tests/test-results/screenshots
```
Comment (nawatts) [inline .github/workflows/run-integration-tests.yaml:131]: Just checking, is this capturing the logs as well as the screenshots?

File: .github/workflows/run-integration-tests.yaml [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    inputs:
+
+      ##
+      ## Configurations:
+      ##
+
+      test_url:
+        required: true
+        type: string
+        description: "URL of terraui instance to test against, required"
```
Comment (nawatts) [inline .github/workflows/run-integration-tests.yaml:47]: Is the intent to usually run these tests against dev, alpha, staging? Or BEEs? If the former, it would be nice to derive `test_url` from `environment` instead of having to specify two parameters that must match.

Comment (jyang-broad) [reply .github/workflows/run-integration-tests.yaml:47]: The intent here is to have this work all 4 scenarios. I would like to give teams the options to run against BEEs in PR environments to cut down on post-PR errors. 

Having BEEs as an option I think throws a wrench in the works for deriving the URL from the env. I added an example call of what client GHAs would look like to call the workflow.

File: .github/workflows/run-integration-tests.yaml (hunk 1 above)
Comment (jyang-broad) [reply .github/workflows/run-integration-tests.yaml:116]: I did this setup specifically because it pulls from current vault secrets that I imagine circleci also uses, it would require duplicating the secrets themselves and reformatting them to make it work, and then we'd end up w/ 2 secrets in vault for the same data that would have to be managed and maintained.

tl;dr: It would make the CI cleaner, but would make the vault setup much worse.

File: .github/workflows/run-integration-tests.yaml (hunk 2 above)
Comment (jyang-broad) [reply .github/workflows/run-integration-tests.yaml:131]: I see a bunch of logs in the /tmp folder, but I think it looks correct

example artifact if you want to look yourself
https://github.com/broadinstitute/monkey/suites/14146347486/artifacts/792371713

log screenshot example:
Screenshot 2023-07-10 at 11 26 08 AM

Comment (nawatts) [reply .github/workflows/run-integration-tests.yaml:131]: Nice. Yeah, that's the type of log I was looking for.

File: .github/workflows/run-integration-tests.yaml (hunk 1 above)
Comment (nawatts) [reply .github/workflows/run-integration-tests.yaml:116]: Ah, didn't realize these were synced from Vault. I think CircleCI's secrets are probably manually copied and pasted, so if this is automated, then this is much better.

Comment (jyang-broad) [reply .github/workflows/run-integration-tests.yaml:116]: Yeah, I'm working w/ the expectation that any repo that tries to call the test would be responsible for pulling the secrets into github via terraform (example below)

https://github.com/broadinstitute/terraform-ap-deployments/blob/df648b0383988781a5299e9066e0d0342f59a676/github/tfvars/broadinstitute-monkey.tfvars#L23

File: .github/workflows/run-integration-tests.yaml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+# While this workflow makes use of secrets, this workflow DOES NOT pull secrets itself and expects upstream callers to
+# provide the secrets via the github secrets functionality.
+#
+# The caller repository must have secrets set up for (Workload Identity Federation is not available for this suite atm)
+#        LYLE_CLIENT_EMAIL
+#        LYLE_PRIVATE_KEY
+#        FIRECLOUD_ACCOUNT_CLIENT_EMAIL
+#        FIRECLOUD_ACCOUNT_PRIVATE_KEY
+# Instruction for getting these secrets securely from vault to github secrets can be found here:
+# https://docs.google.com/document/d/1JbjV4xjAlSOuZY-2bInatl4av3M-y_LmHQkLYyISYns/edit?usp=sharing
```
Comment (nawatts) [inline .github/workflows/run-integration-tests.yaml:13]: Any reason not to add these secrets to the Terra UI repository? I imagine the eventual goal is to use GitHub Actions to run tests on PRs/commits as well.

Comment (jyang-broad) [reply .github/workflows/run-integration-tests.yaml:13]: We'll have to add them for terra-ui to call this workflow from it's own CI, but technically we don't want to make the secrets available by default in the workflow_call because since this is a public repo, this workflow_call is publicly available for anyone to run.

Comment (nawatts) [reply .github/workflows/run-integration-tests.yaml:13]: Ah, I didn't understand how `workflow_call` works (I was thinking of `workflow_dispatch`).

**Review by nawatts — APPROVED** [review]
Thanks for the detailed documentation about how to use this.

File: .github/workflows/run-integration-tests.yaml (hunk 4 above)
Comment (jyang-broad) [reply .github/workflows/run-integration-tests.yaml:13]: 👍 yeah they're similarly named. The overall difference is the the CI run in the caller context, so terra-ui wouldn't be seeing all the CI runs of other repos. basically it's like importing a library.

**Review by Shakespeared — APPROVED** [review]
LGTM, comments all make sense!
